### ISSUE TITLE
App engine GCS build artifacts; comments and constraints

### ISSUE THREAD

**Issue by bsmithyman** [spinnaker/spinnaker#1933]
Hi All,

Following up from a discussion with @danielpeach on Slack, I wanted to provide some feedback and comments on the existing App Engine integration and the planned feature set from the [Q3 2017 Roadmap](https://www.spinnaker.io/community/releases/roadmap/#2017-q3):
- _GCP: Support for fetch/push artifacts with GCS_
- _App Engine: Deployment artifacts from GCS_

I have a bunch of comments... not sure if they will all unpack well into one issue. We are not looking for any particular features, etc. but rather wanted to reach out and contribute some context / comments from our transition to Spinnaker. Our current flows are deploying to GAE through a custom set of scripts on Travis CI (exactly the sort thing we are hoping Spinnaker will move us away from), so I suspect our experiences may be representative for a lot of smaller orgs making this transition.

My understanding is that the next step is supporting build artifacts in a more generalized way through an integration with GCS (i.e., Google Cloud Storage) buckets. I believe the intent is for this to replace / augment the current Git repository integration, but to enable:

1. triggering builds from cloud storage events
2. shipping assets that are not well-suited to being stored directly in Git, through direct upload to the storage buckets from e.g. CI jobs

The rest of this summarizes some thoughts / comments that have come across our plate as we work through gearing up with Spinnaker at my org.

#### Context

The current version of the App Engine **Deploy** step pulls dependencies from a Git repository, and the `gcloud app deploy` command is run from within that checkout. There is provision for multiple kinds of credentials (fixed HTTPS user/pass, SSH). However, there is no hook for additional computation to be done on the deploy, so the contents of the Git repository must be ready to go. This assumes:
- Configuration is inferred from the environment (DB, config management server, etc.) _or_ can be encoded in the `app.yaml` (etc.) in Spinnaker
- Deployed assets do not require pre-compiling (viz., JS, Polymer assets for frontend apps)
- Dependencies are inlined / do not need to be fetched / sourced during a build step (custom code)

I think it is entirely reasonable for scriptability / complex build processes to be out of scope for the Spinnaker / App Engine integration, but it is distinct from some of the other providers that support a bake and deploy flow.

In comparison, our existing CI/CD workflow terminates at the App Engine deploy, but involves collecting versioned packages from public and private sources, running through a lengthy compilation / minification phase for frontend assets, integration testing, configuration parsing. To some extent, this can be simplified away with versioned releases, CDN'd & separated frontend assets, etc. It's a big piece of work for us though (WIP), and probably for others.

#### Issues with just the Git integration

Without a major effort to restructure, we need to ship a "compiled" environment to App Engine. We currently do this via `gcloud app deploy` at the end of our CI workflow. At present, there is no really straightforward way to capture the output of that and "pickle" it for later unpacking on Spinnaker. This is the goal of the GCP upload option (if I am understanding the roadmap correctly).

![image](https://user-images.githubusercontent.com/4269565/29794519-51d35158-8c16-11e7-917f-a67fc3565d56.png)

In order to integrate with Spinnaker in the current system, we can set up an intermediary Git server just for compiled assets. In the CI `after_success:` stage, we go through the "`git init .`; `git add .`; `git commit ...`; `git push -f ...`" flow, and specify the target of that action as the "source" for the deploy stage. Some interesting issues that arise when we try to do this:
- Files are the results of compile steps, so they do not share history; therefore it has to be a "force push"
- The commit hashes seen by Spinnaker do not match up with VCS commits
- Branch targeting in Spinnaker is not aware of commit / build numbers
    - NB: It may be possible to use CI build numbers to specify a branch, unpacking commit and branch state into the branch name on the intermediate server

Depending on how builds work out / what steps are present before the Deploy, it is also possible for the latest "force push" to not be the present execution of the pipeline, and for things to get out of sync.

#### Suggestions for GCS integration

Probably needed:
- A way of matching GCS bucket blobs to a particular build number / branch+commit
    - Supports queueing, reordering, matching with VCS commits, etc
- Triggering pipelines on storage events
    - It sounds like pub/sub integration is planned?

Nice to have:
- Support within `gcloud app ...` to create a packaged build automatically, with the existing logic for collecting files, ignoring paths
    - Maybe an alternate set of verb to `deploy`, something like "`gcloud app package`", with the ability to replay in the deploy later

#### Really surprising and frustrating

There is no easy way to use GCP Cloud Source Repos with Spinnaker
    - No support from the Source Repos side for long-lasting credentials (e.g., SSH access)
    - No support in Spinnaker for using `gcloud` service accounts to access Git
    - Exception: I am guessing this might work with Application Default credentials in some circumstances?
Tighter integration here would be great, since it avoids multiple sets of credentials and keeps everything under the same billing / RBAC.

#### tl;dr
- Existing Git flow works well for ideal apps, not so well if there are major build steps between source and deploy
- Lots of opportunities for the GCS flow to keep a solution inside the Google sphere of influence and make it more seamless
- We really want to use this, but there is some friction... how can we help?

**Comment by danielpeach** [comment]
Thanks @bsmithyman - this is all great. These sorts of problems seem to be the main friction point for people adopting Spinnaker for App Engine.

In general, I don't think Spinnaker should be responsible for building your App Engine app (apart from running `gcloud app deploy`). But I do think the path from built app -> `gcloud app deploy` in Spinnaker could be a lot cleaner.

A few thoughts:

- We've thought about supporting deploying Docker container images to App Engine from Spinnaker (this is already supported from `gcloud`). Just building a container image might end up being easier than anything we come up with.

- There should be some (probably not pretty) way of correlating the app you build (and commit in a separate VCS server) with the commits in your source code. Can you avoid the force push by always pushing to a new branch? Spinnaker could pick up on the branch by matching a regex - you could always be pushing to some new `release-*` branch.

- I haven't looked into supporting GCP Cloud Source Repositories for Spinnaker, but if that's something that would help you along I'd be happy to help scope out the work or advise with the implementation.

- This just occurred to me now, but once the GCS work is done, I am wondering if we could come up with something nice in [Halyard](https://github.com/spinnaker/halyard) to smooth everything over so that you don't have to worry about the GCS logistics and tracability.

e.g., at the end of your build where you would ordinarily run `gcloud app deploy`, you could instead run something like

```
hal app deploy --pipeline PIPELINE_NAME
```

**Comment by bsmithyman** [comment]
Thanks @danielpeach,

Right, that all agrees pretty well with my thinking. I think in our particular case one of the things we need to do is complete our migration to microservices... that means potentially far less baking before the deploy and therefore a better fit with the Spinnaker solution that is designed around that kind of use case.

Re: Docker container
I actually think this is a reasonably good flow already for Flex environment customers. We're limited a bit in that much of our stack is running on GAE Standard, which is necessary if one wants to use NDB (GAE's Python object mapper onto Cloud Datastore). The friction points for getting things to Flex are probably a lot lower because the Flex containers are already able to pull arbitrary dependencies during the build... so, they potentially don't require baking in the same way. That said, we're actually doing our builds and deploys within Docker (lets us run the same code on dev. workstations as on Travis), such that the deploy step is more or less `docker run ... gcloud app deploy`. So, conceivably we can also look at doing a run-around that executes `gcloud` in the Docker image instead of the GAE deploy step. Anyway, that is quite specific to our case; probably not as useful in the general context.

Re: Git and Cloud Source Repositories
Yeah, I am not too concerned in our case about using the repos... we would probably jump to the GCS / direct deploy option if/when it exists. We were able to roll our own midspan git server in pretty short order as a stopover point. I think that there is probably also something that we can do with branch names that encode commit tags; the `release-*` scheme you suggested. The pain point with Cloud Source Repositories is that they seem to require that the machine interacting with them has `gcloud` credentials available (and the ability to talk to the `credential-gcloud.sh` Git credentials helper). I am wondering if there is a straightforward way to get Igor to use those credentials when it looks to Git servers (I assume that is what is doing the pull)? I have not looked into the code closely, but I'll try to dive in sometime when I get a free moment. Anyway, not a big deal for us right now, but just a surprising point of friction in an otherwise pretty complete offering.

The `hal app deploy ...` step is quite interesting; easy is always good. That might be a good way to do more or less what I was suggesting (pickle a `gcloud app deploy` step into an artifact) without adding Spinnaker-specific material to the `gcloud` interface.

**Comment by lwander** [comment]
This is implemented as of 1.7
